File: client/src/main/scala/stasis/client/compression/Compression.scala
package stasis.client.compression

import java.nio.file.Path

import stasis.client.model.EntityMetadata
import stasis.client.model.SourceEntity
import stasis.client.model.TargetEntity

trait Compression {
  def defaultCompression: Encoder with Decoder

  def disabledExtensions: Set[String]

  def algorithmFor(entity: Path): String =
    if (compressionAllowedFor(entity)) {
      defaultCompression.name
    } else {
      Identity.name
    }

  def encoderFor(entity: SourceEntity): Encoder =
    compressionFor(entity.currentMetadata)

  def decoderFor(entity: TargetEntity): Decoder =
    compressionFor(entity.existingMetadata)

  private def compressionAllowedFor(entity: Path): Boolean =
    !disabledExtensions.exists(extension => entity.toString.endsWith(s".$extension"))

  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
  private def compressionFor(entity: EntityMetadata): Encoder with Decoder =
    entity match {
      case file: EntityMetadata.File =>
        Compression.fromString(file.compression)

      case directory: EntityMetadata.Directory =>
        throw new IllegalArgumentException(
          s"Expected metadata for file but directory metadata for [${directory.path.toString}] provided"
        )
    }
}

object Compression {
  def apply(
    withDefaultCompression: String,
    withDisabledExtensions: String
  ): Compression =
    new Compression {
      override val defaultCompression: Encoder with Decoder =
        fromString(compression = withDefaultCompression)

      override val disabledExtensions: Set[String] = withDisabledExtensions
        .split(",")
        .map(_.trim)
        .filter(_.nonEmpty)
        .distinct
        .toSet
    }

  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
  def fromString(compression: String): Encoder with Decoder =
    compression.toLowerCase match {
      case Deflate.name  => Deflate
      case Gzip.name     => Gzip
      case Identity.name => Identity
      case other         => throw new IllegalArgumentException(s"Unexpected compression provided: [$other]")
    }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/compression/Compression.kt
package stasis.client_android.lib.compression

import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.SourceEntity
import stasis.client_android.lib.model.TargetEntity
import java.nio.file.Path
import java.util.Locale

interface Compression {
    fun defaultCompression(): Compressor

    fun disabledExtensions(): Set<String>

    fun algorithmFor(entity: Path): String =
        if (compressionAllowedFor(entity)) {
            defaultCompression().name()
        } else {
            Identity.name()
        }

    fun encoderFor(entity: SourceEntity): Encoder =
        compressionFor(entity.currentMetadata)

    fun decoderFor(entity: TargetEntity): Decoder =
        compressionFor(entity.existingMetadata)

    private fun compressionAllowedFor(entity: Path): Boolean =
        !disabledExtensions().any { extension -> entity.toString().endsWith(".$extension") }

    private fun compressionFor(entity: EntityMetadata): Compressor =
        when (entity) {
            is EntityMetadata.File -> fromString(entity.compression)
            is EntityMetadata.Directory -> throw IllegalArgumentException(
                "Expected metadata for file but directory metadata for [${entity.path}] provided"
            )
        }

    companion object {
        operator fun invoke(
            withDefaultCompression: String,
            withDisabledExtensions: String
        ): Compression = invoke(
            withDefaultCompression = fromString(compression = withDefaultCompression),
            withDisabledExtensions = withDisabledExtensions
                .split(",")
                .map { it.trim() }
                .filter { it.isNotEmpty() }
                .distinct()
                .toSet()
        )

        operator fun invoke(
            withDefaultCompression: Compressor,
            withDisabledExtensions: Set<String>
        ): Compression = object : Compression {
            override fun defaultCompression(): Compressor = withDefaultCompression
            override fun disabledExtensions(): Set<String> = withDisabledExtensions
        }

        fun fromString(compression: String): Compressor =
            when (compression.lowercase(Locale.getDefault())) {
                Deflate.name() -> Deflate
                Gzip.name() -> Gzip
                Identity.name() -> Identity
                else -> throw IllegalArgumentException("Unsupported compression provided: [$compression]")
            }
    }
}
